### PR TITLE
chore(index): fork `index_vec` crate.

### PR DESCRIPTION
**Pull request by rzvxa**
related to #3086

### CODE REVIEW

**Comment by rzvxa** [comment]
* **#3092** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3092?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rzvxa and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/oxc_index/src/idxslice.rs [hunk 1]
```diff
@@ -0,0 +1,847 @@
...
+pub struct IndexSlice<I: Idx, T: ?Sized> {
+    _marker: PhantomData<fn(&I)>,
+    pub raw: T,
+}
+
+// TODO: we may want to remove this somehow. Won't work in upcoming rust versions.
+#[allow(unsafe_code, suspicious_auto_trait_impls)]
+// SAFETY: Whether `IndexSlice` is `Send` depends only on the data,
+// not the phantom data.
+unsafe impl<I: Idx, T> Send for IndexSlice<I, [T]> where T: Send {}
```
Comment (rzvxa) [inline crates/oxc_index/src/idxslice.rs:49]: This will break in the short term future. https://github.com/rust-lang/rust/issues/93367

Comment (rzvxa) [reply crates/oxc_index/src/idxslice.rs:49]: I'll create an issue for it after the merge so we can keep track of it.

**Comment by rzvxa** [comment]
I intentionally didn't add a license file from the original crate so you can review and suggest how are we going to pay attribute to the author.

**Comment by Boshen** [comment]
> I intentionally didn't add a license file from the original crate so you can review and suggest how are we going to pay attribute to the author.

I put licenses in https://github.com/oxc-project/oxc/blob/main/THIRD-PARTY-LICENSE

And also mention this is a fork from index_vec on the top of the lib.rs file.

**Review by Boshen — APPROVED** [review]

**Comment by Boshen** [comment]
Feel free to merge.

**Review by Boshen — APPROVED** [review]

**Comment by overlookmotel** [comment]
One bit of feedback: You can avoid compiling the `serde` dependency when the `serialize` feature isn't enabled:

In `Cargo.toml`:

```diff
[dependencies]
- serde = { workspace = true }
+ serde = { workspace = true, optional = true }

[features]
- serialize = []
+ serialize = ["dep:serde"]
```

**Comment by rzvxa** [comment]
@overlookmotel Oh thanks, I forgot to add the serde as optional. I'll fix it.

**Pull request merged**